src/unity/lib/extensions/model_base.cpp

```diff
@@ -0,0 +1,126 @@
+/* Copyright © 2017 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+ */
+#include <unity/lib/extensions/model_base.hpp>
+
+namespace turi {
+
+model_base::~model_base() = default;
+
+void model_base::save_impl(oarchive& oarc) const {}
+
+void model_base::load_version(iarchive& iarc, size_t version) {}
+
+size_t model_base::get_version() const {
+  return 0;
+}
+
+std::map<std::string, std::vector<std::string>> model_base::list_functions() {
+  perform_registration();
+  return m_function_args;
+}
+
+std::vector<std::string> model_base::list_get_properties() {
+  perform_registration();
+  std::vector<std::string> ret;
+  for(const auto& i: m_get_property_list) ret.push_back(i.first);
+  return ret;
+}
+
+std::vector<std::string> model_base::list_set_properties() {
+  perform_registration();
+  std::vector<std::string> ret;
+  for(const auto& i: m_set_property_list) ret.push_back(i.first);
+  return ret;
+}
+
+variant_type model_base::call_function(std::string function, 
+				       variant_map_type argument) {
+  perform_registration();
+  if (m_function_list.count(function)) {
+    // fill in default args if any
+    if (m_function_default_args.count(function)) {
```

Comment:
I know this is just moved code, but it would be good to avoid multiple lookups here.  

I think what I'm going to do is open an issue to optimize these function calls as they are part of a pretty hot path, but let's not do that right now.